odoo/addons/base/models/ir_model.py

```diff
@@ -1227,6 +1234,43 @@ def write(self, vals):
 
         return result
 
+    @api.model
+    def __update_reflected_field(self, table, field, value, selection_value):
+        query = 'UPDATE "{table}" SET "{field}"=%s WHERE "{field}"=%s'.format(
+            table=table, field=field,
+        )
+        self.env.cr.execute(query, (value, selection_value,))
+
+    @api.model
+    def __delete_reflected_records(self, table, field, selection_value):
+        query = 'DELETE FROM "{table}" WHERE "{field}"=%s'.format(
+            table=table, field=field,
+        )
+        self.env.cr.execute(query, (selection_value,))
+
+    def _handle_ondelete(self):
+        for selection in self.filtered('field_id.store'):
+            args = (self.env[selection.field_id.model]._table, selection.field_id.name,)
+            value = selection.value
+
+            Model = self.env[selection.field_id.model]
+            field = Model._fields[selection.field_id.name]
+            # TODO: simplify and maybe convert None to 'null' in _setup_regular_base()
+            if not selection.ondelete or selection.ondelete == 'null':
+                self.__update_reflected_field(*args, None, value)
+            elif selection.ondelete == 'default':
+                default_value = field.default(self.env[selection.field_id.model])
+                self.__update_reflected_field(*args, default_value, value)
+            elif selection.ondelete == 'callable':
+                recs = Model.search([(field.name, '=', value)])
+                field.ondelete[value](recs)
+            elif selection.ondelete == 'cascade':
+                self.__delete_reflected_records(*args, value)
+            else:
+                # this is okay as _setup_regular_base has already verified that the input is valid
+                _, literal = selection.ondelete.split(maxsplit=1)
+                self.__update_reflected_field(*args, literal, value)
```

Comment:
I no longer remember the argument for doing this in plain SQL.
Should we expect business logic to apply?
Should we expect field recomputations?

Reply:
There was none, I guess I could do it through the ORM just as well, but I wonder if this ondelete policy is even necessary anymore with callbacks being implemented? You save a few characters of code at most and the API would be cleaner without it

Reply:
That's what I thought.  The following
```python
ondelete = {
    'foo': 'null',
    'bar': 'cascade',
    'baz': 'default',
    'quux': 'set other',
}
```
is mostly (modulo business logic) equivalent to
```python
ondelete = {
    'foo': lambda recs: recs.write({'stuff': False}),
    'bar': lambda recs: recs.unlink(),
    'baz': lambda recs: recs.write({'stuff': recs._fields['stuff'].default(recs)}),
    'quux': lambda recs: recs.write({'stuff': 'other'}),
}
```

Reply:
Note that an SQL UPDATE will be more efficient than a write(), in case there are many records to update.